core/src/queue.rs

```diff
@@ -322,46 +294,64 @@ impl Queue {
         });
-        let txs_from_waiting_buffer = core::iter::from_fn(|| {
-            self.pop_from_signature_buffer(
-                &mut seen_waiting_buffer,
-                wsv,
-                &mut expired_transactions_waiting_buffer,
-            )
-        });
 
         let transactions_hashes: HashSet<HashOf<VersionedSignedTransaction>> = transactions
             .iter()
             .map(VersionedAcceptedTransaction::hash)
             .collect();
         let txs = txs_from_queue
-            .riffle(txs_from_waiting_buffer)
             .filter(|tx| !transactions_hashes.contains(&tx.hash()))
             .take(max_txs_in_block - transactions.len());
         transactions.extend(txs);
 
-        [
-            (seen_queue, &self.queue, expired_transactions_queue),
-            (
-                seen_waiting_buffer,
-                &self.signature_buffer,
-                expired_transactions_waiting_buffer,
-            ),
-        ]
-        .into_iter()
-        .for_each(|(seen, queue, expired_txs)| {
-            seen.into_iter()
-                .try_for_each(|hash| queue.push(hash))
-                .expect("Exceeded the number of transactions pending");
-            expired_transactions.extend(expired_txs);
-        })
+        seen_queue
+            .into_iter()
+            .try_for_each(|hash| self.queue.push(hash))
+            .expect("Exceeded the number of transactions pending");
+        expired_transactions.extend(expired_transactions_queue);
+    }
+
+    /// Check that user adhere for maximum transaction per user limit and increase his transaction count.
+    ///
+    /// Check for limit and increment is done in one go in order to prevent situation
+    /// when multiple transactions from the same user are trying to increase counter at the same time.
+    fn check_and_increase_per_user_tx_count(&self, account_id: &AccountId) -> Result<(), Error> {
+        match self.txs_per_user.entry(account_id.clone()) {
+            Entry::Vacant(vacant) => {
+                vacant.insert(1);
+            }
+            Entry::Occupied(mut occupied) => {
+                let txs = *occupied.get();
+                if txs >= self.max_txs_per_user {
+                    warn!(
+                        max_txs_per_user = self.max_txs_per_user,
+                        %account_id,
+                        "Account reached maximum allowed number of transactions in the queue per user"
+                    );
+                    return Err(Error::MaximumTransactionsPerUser);
+                }
+                *occupied.get_mut() += 1;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn decrease_per_user_tx_count(&self, account_id: &AccountId) {
+        let Entry::Occupied(mut occupied) = self.txs_per_user
+            .entry(account_id.clone()) else { panic!("Call to decrease always should be paired with increase count. This is a bug.") };
```

Comment:
```suggestion
        let mut occupied = self.txs_per_user[account_id.clone()];
```

should be as good + removes runtime check

Reply:
Unfortunately `DashMap` doesn't implement `IndexMut` trait.